src/main/java/com/tagtraum/perf/gcviewer/imp/DataReaderShenandoah.java

```diff
@@ -99,50 +108,14 @@ public GCModel read() throws IOException {
-                    Double.parseDouble(noHeapMatcher.group(NO_HEAP_TIMESTAMP).replace(",", ".")));
+            AbstractGCEvent.Type type = AbstractGCEvent.Type.lookup(noHeapMatcher.group(NO_HEAP_EVENT_NAME));
+            event.setType(type);
+            setPauseAndTimestamp(event, noHeapMatcher.group(NO_HEAP_DURATION), noHeapMatcher.group(NO_HEAP_TIMESTAMP));
         } else if (withHeapMatcher.find()) {
-            // Concurrent events
-            if (line.contains("Concurrent")) {
-                event = new ConcurrentGCEvent();
-                if (line.contains("Concurrent marking")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_CONCURRENT_CONC_MARK);
-                } else if (line.contains("Concurrent evacuation")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_CONCURRENT_CONC_EVAC);
-                } else if (line.contains("Concurrent update references")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_CONCURRENT_CONC_UPDATE_REFS);
-                } else if (line.contains("Concurrent reset bitmaps")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_CONCURRENT_CONC_RESET_BITMAPS);
-                } else if (line.contains("Concurrent precleaning")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_CONCURRENT_PRECLEANING);
-                }
-            }
-            // STW events
-            else {
-                event = new GCEvent();
-                if (line.contains("Final Mark")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_STW_FINAL_MARK);
-                } else if (line.contains("Pause Full (Allocation Failure)")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_STW_ALLOC_FAILURE);
-                } else if (line.contains("Pause Final Update Refs")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_STW_FINAL_UPDATE_REFS);
-                } else if (line.contains("Pause Full (System.gc())")) {
-                    event.setType(AbstractGCEvent.Type.SHEN_STW_SYSTEM_GC);
-                } else {
-                    getLogger().warning("Failed to match line with heap info: " + line);
-                }
-            }
-            setPauseAndTimestamp(event,
-                    Double.parseDouble(withHeapMatcher.group(WITH_HEAP_DURATION).replace(",", ".")),
-                    Double.parseDouble(withHeapMatcher.group(WITH_HEAP_TIMESTAMP).replace(",", ".")));
+            event = line.contains("Concurrent") ? new ConcurrentGCEvent() : new GCEvent();
+            AbstractGCEvent.Type type = AbstractGCEvent.Type.lookup(withHeapMatcher.group(WITH_HEAP_EVENT_NAME));
+            event.setType(type);
+            setPauseAndTimestamp(event, withHeapMatcher.group(WITH_HEAP_DURATION), withHeapMatcher.group(WITH_HEAP_TIMESTAMP));
             addHeapDetailsToEvent(event, withHeapMatcher.group(WITH_HEAP_MEMORY));
         } else {
             getLogger().warning("Found line that has no match:" + line);
```

Comment:
Usually, I add the line number to the log warnings, because this makes finding the offending line in the log easier. You might be able to use in.getLineNumber() (might not work here, because of the usage of the streams - I don't have any experience, there).

If it is difficult to add the line number - no problem. The warning also helps without the line number.